datafusion/physical-plan/src/repartition/on_demand_repartition.rs

```diff
@@ -0,0 +1,1177 @@
+                for (_, tx) in txs {
+                    tx.send(None).await.ok();
+                }
+            }
+        }
+    }
+}
+
+/// This struct converts a receiver to a stream.
+/// Receiver receives data on an SPSC channel.
+struct OnDemandPerPartitionStream {
+    /// Schema wrapped by Arc
+    schema: SchemaRef,
+
+    /// channel containing the repartitioned batches
+    receiver: DistributionReceiver<MaybeBatch>,
+
+    /// Handle to ensure background tasks are killed when no longer needed.
+    _drop_helper: Arc<Vec<SpawnedTask<()>>>,
+
+    /// Memory reservation.
+    reservation: SharedMemoryReservation,
+
+    /// Sender to send partititon number to the receiver
+    sender: Sender<usize>,
+
+    /// Partition number
+    partition: usize,
+}
+
+impl Stream for OnDemandPerPartitionStream {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        // TODO: To be removed
+        debug!(
+            "On Demand Repartition per partition poll {}",
+            self.partition
+        );
+        if !self.sender.is_closed() {
+            self.sender.send_blocking(self.partition).map_err(|e| {
+                internal_datafusion_err!(
+                    "Error sending partition number to input partitions: {}",
+                    e
+                )
+            })?;
+        }
+
+        match ready!(self.receiver.recv().poll_unpin(cx)) {
+            Some(Some(v)) => {
+                if let Ok(batch) = &v {
+                    self.reservation
+                        .lock()
+                        .shrink(batch.get_array_memory_size());
+                }
+
+                Poll::Ready(Some(v))
+            }
+            Some(None) => {
+                // Input partition has finished sending batches
+                Poll::Ready(None)
+            }
+            None => Poll::Ready(None),
+        }
+    }
+}
+
+impl RecordBatchStream for OnDemandPerPartitionStream {
+    /// Get the schema
+    fn schema(&self) -> SchemaRef {
+        Arc::clone(&self.schema)
+    }
+}
+
+struct OnDemandRepartitionStream {
+    /// Number of input partitions that will be sending batches to this output channel
+    num_input_partitions: usize,
+
+    /// Number of input partitions that have finished sending batches to this output channel
+    num_input_partitions_processed: usize,
+
+    /// Schema wrapped by Arc
+    schema: SchemaRef,
+
+    /// channel containing the repartitioned batches
+    input: DistributionReceiver<MaybeBatch>,
+
+    /// Handle to ensure background tasks are killed when no longer needed.
+    _drop_helper: Arc<Vec<SpawnedTask<()>>>,
+
+    /// Memory reservation.
+    reservation: SharedMemoryReservation,
+
+    /// Sender for the output partition
+    sender: Sender<usize>,
+
+    /// Partition number
+    partition: usize,
+
+    /// Sender state
+    is_requested: bool,
```

Comment:
If the receiver is waiting for the data, but it has sent the partition ID previously. `is_requested` is `TRUE`